spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/metrics/webservices/client/WebServiceTemplateExchangeTags.java

```diff
@@ -0,0 +1,131 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.boot.actuate.metrics.webservices.client;
+
+import io.micrometer.core.instrument.Tag;
+
+import org.springframework.util.StringUtils;
+import org.springframework.ws.WebServiceMessage;
+import org.springframework.ws.client.WebServiceClientException;
+import org.springframework.ws.client.WebServiceIOException;
+import org.springframework.ws.context.MessageContext;
+import org.springframework.ws.soap.SoapMessage;
+import org.springframework.ws.transport.WebServiceConnection;
+import org.springframework.ws.transport.context.TransportContext;
+import org.springframework.ws.transport.context.TransportContextHolder;
+
+/**
+ * Factory methods for creating {@link Tag Tags} related to a soap request-response
+ * exchange performed by a {@link org.springframework.ws.client.core.WebServiceTemplate}.
+ *
+ * @author Dmytro Nosan
+ * @since 2.1.0
+ **/
+public abstract class WebServiceTemplateExchangeTags {
+
+	private static final Tag OK = Tag.of("status", "OK");
+
+	private static final Tag IO_ERROR = Tag.of("status", "IO_ERROR");
+
+	private static final Tag CLIENT_ERROR = Tag.of("status", "CLIENT_ERROR");
+
+	private static final Tag SERVER_ERROR = Tag.of("status", "SERVER_ERROR");
+
+	private static final Tag NO_URI = Tag.of("uri", "NONE");
+
+	private static final Tag NO_ACTION = Tag.of("action", "NONE");
+
+	/**
+	 * Creates a {@code Tag(uri)} using {@link TransportContext#getConnection()}.
+	 * @return the uri tag
+	 */
+	public static Tag uri() {
+		String uri = getUri();
+		return (!StringUtils.hasText(uri) ? NO_URI : Tag.of("uri", uri));
+	}
+
+	/**
+	 * Creates a {@code Tag(action)} using {@link SoapMessage#getSoapAction()}.
+	 * @param messageContext Context of the SOAP message.
+	 * @return the action tag
+	 */
+	public static Tag action(MessageContext messageContext) {
+		String action = getAction(messageContext);
+		return (!StringUtils.hasText(action) ? NO_ACTION : Tag.of("action", action));
+	}
+
+	/**
+	 * Creates a {@code Tag(status)} based on the exception.
+	 * @param exception the exception
+	 * @return the status tag
+	 */
+	public static Tag status(Exception exception) {
+		if (exception == null) {
+			return OK;
+		}
+		if (exception instanceof WebServiceIOException) {
+			return IO_ERROR;
+		}
+		if (exception instanceof WebServiceClientException) {
+			return CLIENT_ERROR;
+		}
+
+		return SERVER_ERROR;
+	}
+
+	private static String getUri() {
```

Comment:
Does this return a parameterized uri. For example `/foo/{id}` instead of `/foo/1`?

Reply:
`WebServiceTemplate` doesn't have `uriTemplates`.

Reply:
What can we do then to limit tag cardinality?

Reply:
Should we do?  With `RestTemplate` it could be an issue with `uriTemplates`, here we don't have it.

Reply:
This speaks to my ignorance of SOAP I think. Are you saying SOAP URIs are always fixed and never have path variables in them?

Reply:
There are 2 often cases.
 -  Fixed `URI` -  different `SoapAction`
- Different `URI` , Fixed `SoapAction` (e.g. "")

for the `RestTemplate` it could be a case when people use `URI + variables` instead of `uriVariables` map, thus instead of one `tag {id}` will be N `tags`.

**P.S.** We can easily add `MeterFilter` just to make sure everything is ok, as `Spring Boot` did for `RestTemplate`.